### PR TITLE
CRLP: Rollup Grouping Refactor to fix bug with YearsAgo rollup not working

### PR DESCRIPTION
**Pull request by force2b**
Refactor the `CRLP_Rollup` class and more specifically the grouping logic to:
* Fix a bug with how Rollup records were grouped especially when DaysAgo, YearsBack or Count operations are involved; especially with Single Result operations (First/Last/Largest/Smallest).
* Reduce the heap size used by CRLP_Rollup instances

# Critical Changes

# Changes

# Issues Closed

* Fixes #4232 

# New Metadata

# Deleted Metadata

### CODE REVIEW

**Comment by force2b** [comment]
**lurch: attach W-030691

**Comment by LurchTheButler** [comment]
Tracking <a href="https://foundation.my.salesforce.com/a2x1E000007fQhaQAE" target="_blank">W-030691</a>

File: src/classes/CRLP_Batch_Base_Skew.cls [hunk 1]
```diff
@@ -313,10 +314,10 @@ public abstract class CRLP_Batch_Base_Skew extends CRLP_Batch_Base {
         if (detailObjectType == Opportunity.SObjectType) {
             addlFilter = CRLP_Query_SEL.buildCommonQueryFilterOnOpportunity(this.summaryObjectType, detailObjectType,
                     this.statefulCacheOfRollupsToBeProcessed);
-        } else if (this.jobType == CRLP_RollupProcessingOptions.RollupType.ContactSoftCredit) {
+
+        } else if (CRLP_Rollup_SVC.isOppContactRoleSoftCreditRollup(this.jobType)) {
             addlFilter = CRLP_Query_SEL.buildCommonQueryFilterOnOpportunity(Contact.SObjectType, Partial_Soft_Credit__c.SObjectType);
-        } else if (this.jobType == CRLP_RollupProcessingOptions.RollupType.AccountContactSoftCredit) {
-            addlFilter = CRLP_Query_SEL.buildCommonQueryFilterOnOpportunity(Account.SObjectType, Partial_Soft_Credit__c.SObjectType);
```
[the remark this answers is not in the document]
Comment (force2b) [reply src/classes/CRLP_Batch_Base_Skew.cls]: `ContactSoftCredit` and `AccountContactSoftCredit` are both based on the `Partial_Soft_Credit` object. I moved the logic that check for those two options into the `CRLP_Rollup_SVC.isOppContactRoleSoftCreditRollup()` method. The `AccountSoftCredit` rollup type uses the 'Account_Soft_Credit` object. The if for this is on line 321.

File: src/classes/CRLP_Operation_SVC.cls [hunk 2]
```diff
@@ -154,77 +186,127 @@ public class CRLP_Operation_SVC {
...
+        if (rollup.detailObject != objName && rollup.detailObject != oppObjectName) {
+            return;
+        }
+
+        // For each RollupDefinition within this group where the Rollup is a SingleResultType
+        // (First, Last, Largest, Smallest)
+        for (Integer ns = 0, listSize = rollup.rollupsMdt.size(); ns < listSize; ns++) {
+            CRLP_Rollup.Rollupmdt r = rollup.rollupsMdt[ns];
+
+            // if this isn't a single result operation rollup, just skip the loop
```
[the remark this answers is not in the document]
Comment (force2b) [reply src/classes/CRLP_Operation_SVC.cls]: Definitely. Nice catch.

**Comment by mrbelvedere** [comment]
Included in beta release 3.156 (Beta 13)

**Pull request merged**